---
index: gnomad_help  
title: 'SV consequence: loss of function' 
---

## ![](https://placehold.it/15/D43925/000000?text=+) Predicted loss-of-function (pLoF)

### Description

SV resulting in pLoF of a gene are predicted to delete protein-coding sequence, or otherwise truncate the transcript or gene product.  

On average, these variants can be expected to reduce gene function by 50% when in the heterozygous state, and by 100% when in the homozygous state.  

### Disclaimer

This is a relatively broad categorization of pLoF variants across SV classes. Not all pLoF SVs will result in true loss-of-function.  

Just like for SNVs & indels, please [carefully inspect each pLoF SV before interpretation](https://broad.io/gnomad_drugs), particularly in the context of [its exonic overlap and affected transcripts](https://broad.io/tx_annotation).

### Annotation criteria

An SV was assigned a `loss of function` consequence if it met any of the following criteria:

![Predicted loss-of-function (pLoF)](gnomAD_browser.effect_schematics_pLoF.png)    
